qemu/qemu.go

```diff
@@ -2103,6 +2103,54 @@ type Kernel struct {
 	Params string
 }
 
+// FwCfg allows QEMU to pass entries to the guest
+// File and Str are mutually exclusive
+type FwCfg struct {
+	Name string
+	File string
+	Str  string
+}
+
+// Valid returns true if the FwCfg structure is valid and complete.
+func (fwcfg FwCfg) Valid() bool {
+	if fwcfg.Name == "" {
+		return false
+	}
+
+	if fwcfg.File != "" && fwcfg.Str != "" {
+		return false
+	}
+
+	if fwcfg.File == "" && fwcfg.Str == "" {
+		return false
+	}
+
+	return true
+}
+
+// QemuParams returns the qemu parameters built out of the FwCfg object
+func (fwcfg FwCfg) QemuParams(config *Config) []string {
+	var fwcfgParams []string
+	var qemuParams []string
+
+	if config.FwCfg.Name != "" {
+		fwcfgParams = append(fwcfgParams, fmt.Sprintf("name=%s", config.FwCfg.Name))
+
+		if config.FwCfg.File != "" {
+			fwcfgParams = append(fwcfgParams, fmt.Sprintf(",file=%s", config.FwCfg.File))
+		}
+
+		if config.FwCfg.Str != "" {
+			fwcfgParams = append(fwcfgParams, fmt.Sprintf(",string=%s", config.FwCfg.Str))
```

Comment:
Any of the `FwCfg` elements could contain a space presumably, hence shouldn't these formats use `%q` rather than `%s`?

Reply:
Yes you are right.